### PR TITLE
Remove use of alpha initializers

### LINKED ISSUE #67113: Consider deprecate `Initializers` admission controller
As webhook admission has been beta, and we can implement all the function` Initializers` does by webhook admission. So I think it's time to deprecate `Initializers`, eventually remove it.

/cc @caesarxuchao @deads2k @liggitt @sttts

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
* Marks the alpha metadata.Initializers field deprecated and removes all in-tree use of it
* Removes the admissionregistration.k8s.io/v1alpha1 API group, containing only the InitializationConfiguration type
* Removes the `IncludeUninitialized` field from `*Options`
* Removes the alpha Initializers admission plugin
* Removes use of Initializers fields from the cloud-controller-manager persistent volume labeler controller (stubs the needsInitialization() method to return false, and  the `markInitialized()` method to no-op). This controller will either need to work off of a different attribute (like an annotation), or become an admission webhook.

**Which issue(s) this PR fixes**:
Fixes #67113 

**Special notes for your reviewer**:

There are no plans to graduate the initializers feature, so removing it sooner rather than later sends a clearer signal to the community that new features should not be built anticipating its promotion.

**Does this PR introduce a user-facing change?**:
```release-note-action-required
The alpha Initializers feature, `admissionregistration.k8s.io/v1alpha1` API version, `Initializers` admission plugin, and use of the `metadata.initializers` API field have been removed. Discontinue use of the alpha feature and delete any existing `InitializerConfiguration` API objects before upgrading. The `metadata.initializers` field will be removed in a future release.
```

/sig api-machinery
/cc @lavalamp @deads2k @smarterclayton @caesarxuchao

### CODE REVIEW

**Comment by liggitt** [comment]
added to the [sig-api-machinery meeting agenda on 1/16](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.rcsstt379k5w) for discussion

**Comment by smarterclayton** [comment]
Did we want to keep the field so that we don't break downstream and remove it in a future release?

**Comment by liggitt** [comment]
> Did we want to keep the field so that we don't break downstream and remove it in a future release?

Current state:
* feature is disabled by default
* when the feature is disabled, any metadata.initializers sent to the apiserver are ignored and not persisted
* if a local manifest has a `metadata.initializers` field present, kubectl client-side validation will not complain because the field appears in the openapi schema

We can consider keeping the field in metadata and unconditionally clearing it server-side, but I don't see what would make it safer to remove in a future release than it is now

I'd prefer to remove it from the schema to avoid generating misleading clients

**Comment by liggitt** [comment]
@kubernetes/sig-cloud-provider-pr-reviews 
for impact on PV label controller which was built on the alpha initializers feature

**Comment by andrewsykim** [comment]
@liggitt is this targeting v1.14?

**Comment by liggitt** [comment]
updated this to keep the metadata.initializers field (for now), mark it deprecated, and remove all use of it in tree.

can continue to discuss when/how we want to drop the field entirely

File: pkg/controller/cloud/pvlcontroller.go [hunk 1]
```diff
@@ -44,8 +44,6 @@ import (
 	volumeutil "k8s.io/kubernetes/pkg/volume/util"
 )
 
-const initializerName = "pvlabel.kubernetes.io"
```
Comment (andrewsykim) [inline pkg/controller/cloud/pvlcontroller.go]: Still working through alternatives to this. Will loop back when we come up with something, hopefully within the next few days

**Comment by liggitt** [comment]
all CI is green

File: pkg/controller/cloud/pvlcontroller.go (hunk 1 above)
Comment (andrewsykim) [reply pkg/controller/cloud/pvlcontroller.go]: Spoke with @dims offline. Given that there are very few users actually using Initializers for PV labeling we're just going to port any remaining implementations of the PV labeler (OpenStack) into the in-tree PVL admission controller. This should buy us some time until a better solution that doesn't use Initializers is available out-of-tree. PR to update PVL admission controller coming soon.

Comment (dims) [reply pkg/controller/cloud/pvlcontroller.go]: we will have to poll folks on sig-openstack slack and email, but at least we will have an alternative to recommend.

**Comment by liggitt** [comment]
> @liggitt the cloud provider usage of initializer should not block this any more - #73102

great, thanks

**Review by caesarxuchao — COMMENTED** [review]
Only nits. LGTM.

**Comment by liggitt** [comment]
updated

**Comment by caesarxuchao** [comment]
Unit tests for the two modified admission plugin failed.

File: pkg/controller/cloud/pvlcontroller.go [hunk 2]
```diff
@@ -286,38 +282,13 @@ func (pvlc *PersistentVolumeLabelController) updateVolume(vol *v1.PersistentVolu
...
-
-	if len(initializers.Pending) == 0 {
-		return false
-	}
-
-	// There is at least one initializer still pending so check to
-	// see if the PVL is the next in line.
-	return initializers.Pending[0].Name == name
+func needsInitialization(vol *v1.PersistentVolume) bool {
+	// TODO: determine whether initialization is required based on a different attribute,
```
Comment (andrewsykim) [inline pkg/controller/cloud/pvlcontroller.go]: Just a note for reference: this should ideally be done for v1.14 (in a future PR though)

**Comment by liggitt** [comment]
this is ready, local-e2e is red for an unrelated reason

**Comment by liggitt** [comment]
Thanks for the feedback, I'll send an announcement to kubernetes-dev as well

**Comment by akutz** [comment]
Hi @liggitt,

Thank you. I feel like I'm the [Alpha-thorn](https://github.com/kubernetes/kubernetes/pull/66916) in your side :(

FWIW, I used this config bit because of [Kelsey's guide](https://github.com/kelseyhightower/kubernetes-the-hard-way/blob/master/docs/08-bootstrapping-kubernetes-controllers.md#configure-the-kubernetes-api-server). I'll file a PR with his repo as well. 

Thank you Jordan!

-- 
-a

**Comment by simonferquel** [comment]
`IncludeInitialize` on Get and ListOptions also allowed to query for objects pending finalization. How can we do that now ?

**Comment by liggitt** [comment]
> `IncludeInitialize` on Get and ListOptions also allowed to query for objects pending finalization. How can we do that now ?

That is no longer possible. That functionality never graduated from alpha-level in the server, so servers using only beta-level functionality do not have uninitialized objects.

**Comment by simonferquel** [comment]
@liggitt I understand, but IncludeUninitialized also had the effect to be able to list resources that are pending finalization (e.g. those that have a non-nil DeletedTimestamp, but a non-empty list of finalizers). How can we list resources that are in this case ? (my use case is within a custom controller, which uses a watch and respond to this condition)

**Comment by simonferquel** [comment]
(an example of that, is for writing custom finalizers: https://github.com/hossainemruz/k8s-initializer-finalizer-practice/blob/master/pkg/controller/deploymentcontroller.go#L62)

**Comment by liggitt** [comment]
> @liggitt I understand, but IncludeUninitialized also had the effect to be able to list resources that are pending finalization (e.g. those that have a non-nil DeletedTimestamp, but a non-empty list of finalizers).

Objects pending finalization are included regardless of the `IncludeUninitialized` parameter.

**Pull request merged**